drivers/xen/gnttab.c

```diff
@@ -0,0 +1,336 @@
+ *
+ *        Date: July 2006
+ *
+ * Environment: Xen Minimal OS
+ * Description: Simple grant tables implementation. About as stupid as it's
+ *  possible to be and still work.
+ *
+ ****************************************************************************
+ */
+#include <arch/arm64/hypercall.h>
+#include <xen/generic.h>
+#include <xen/gnttab.h>
+#include <xen/public/grant_table.h>
+#include <xen/public/memory.h>
+#include <xen/public/xen.h>
+
+#include <init.h>
+#include <kernel.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(xen_gnttab);
+
+/* Timeout for grant table ops retrying */
+#define GOP_RETRY_DELAY 200
+
+/* NR_GRANT_FRAMES must be less than or equal to that configured in Xen */
+#define NR_GRANT_FRAMES			1
+#define NR_GRANT_ENTRIES \
+	(NR_GRANT_FRAMES * XEN_PAGE_SIZE / sizeof(grant_entry_v1_t))
+
+static struct gnttab {
+	struct k_sem sem;
+	grant_entry_v1_t *table;
+	grant_ref_t gref_list[NR_GRANT_ENTRIES];
+} gnttab;
+
+static grant_ref_t get_free_entry(void)
+{
+	grant_ref_t gref;
+	unsigned int flags;
+
+	k_sem_take(&gnttab.sem, K_FOREVER);
+
+	flags = irq_lock();
+	gref = gnttab.gref_list[0];
+	__ASSERT((gref >= GNTTAB_NR_RESERVED_ENTRIES &&
+		gref < NR_GRANT_ENTRIES), "Invalid gref = %d", gref);
+	gnttab.gref_list[0] = gnttab.gref_list[gref];
+	irq_unlock(flags);
+
+	return gref;
+}
+
+static void put_free_entry(grant_ref_t gref)
+{
+	unsigned int flags;
+
+	flags = irq_lock();
+	gnttab.gref_list[gref] = gnttab.gref_list[0];
+	gnttab.gref_list[0] = gref;
+
+	irq_unlock(flags);
+
+	k_sem_give(&gnttab.sem);
+}
+
+static void gnttab_grant_init(grant_ref_t gref, domid_t domid,
+		unsigned long mfn)
+{
+	gnttab.table[gref].frame = mfn;
+	gnttab.table[gref].domid = domid;
+
+	compiler_barrier();
+}
+
+static void gnttab_grant_permit_access(grant_ref_t gref, domid_t domid,
+		unsigned long mfn, int readonly)
+{
+	gnttab_grant_init(gref, domid, mfn);
+	readonly *= GTF_readonly;
+	gnttab.table[gref].flags = GTF_permit_access | readonly;
+}
+
+grant_ref_t gnttab_grant_access(domid_t domid, unsigned long mfn,
+		int readonly)
+{
+	grant_ref_t gref = get_free_entry();
+
+	gnttab_grant_permit_access(gref, domid, mfn, readonly);
+
+	return gref;
+}
+
+grant_ref_t gnttab_grant_transfer(domid_t domid, unsigned long mfn)
+{
+	grant_ref_t gref = get_free_entry();
+
+	gnttab_grant_init(gref, domid, mfn);
+	gnttab.table[gref].flags = GTF_accept_transfer;
+
+	return gref;
+}
+
+/* Reset flags to zero in order to stop using the grant */
+static int gnttab_reset_flags(grant_ref_t gref)
+{
+	uint16_t flags, nflags;
+	uint16_t *pflags;
+
+	pflags = &gnttab.table[gref].flags;
+	nflags = *pflags;
+
+	do {
+		flags = nflags;
+		if (flags & (GTF_reading | GTF_writing)) {
+			LOG_WRN("gref = %u still in use! (0x%x)\n",
+				gref, flags);
+			return 0;
+		}
+		nflags = synch_cmpxchg(pflags, flags, 0);
+	} while (nflags != flags);
+
+	return 1;
+}
+
+int gnttab_update_grant(grant_ref_t gref, domid_t domid,
```

Comment:
Can you clarify how this function is intended to be used? My comments on this function will depend on this answer, so I will wait before review it.

Reply:
This function was present in original driver and per my understanding it is needed for changing parameters for already taken gref (e.g. change readonly status at some moment). I do not know if it is a right way to do this (or is it even allowed), so I will be happy if you clarify this point.

Reply:
I think you want to only introduce helpers that will be used. In this case, `gnttab_update_grant()` can only be used if the grant-table reference is not used by the other-side.

So the usability of this function looks quite limited to me. Do you need in a driver you will implement? If yes, do you have an example how this will be used? If not, then I would say to drop it until you need it.